docs/getting-started/configuration.md
# Configuration

This document explains the environment variables used to configure your Enviro deployment. All variables are set in the `.env` file at the root of your project. You can use `.env.example` as a template.

## Django Settings

- **DJANGO_SECRET_KEY**  
  Secret key for Django. Change this to a unique, unpredictable value in production.

- **DEFAULT_SCHEME**
  Set to `https` or `http`. This will set the redirects scheme.

- **DEBUG**  
  Set to `True` for development, `False` for production.

- **DJANGO_LOGLEVEL**  
  Logging level for Django. Common values: `info`, `debug`, `warning`, `error`.

- **DJANGO_ALLOWED_HOSTS**  
  Comma-separated list of hostnames/domain names that this Django site can serve.

- **DJANGO_CSRF_TRUSTED_ORIGINS**
  Comma-seperated list of trusted csrf origins that Django can trust. Default `http://localhost:8081`.

## Database Settings

- **DATABASE_ENGINE**  
  Database backend to use. Example: `postgresql_psycopg2`.

- **DATABASE_NAME**  
  Name of the database.

- **DATABASE_USERNAME**  
  Username for the database.

- **DATABASE_PASSWORD**  
  Password for the database user.

- **DATABASE_HOST**  
  Hostname or IP address of the database server.

- **DATABASE_PORT**  
  Port number for the database server.

## Celery Settings

- **CELERY_BROKER_URL**  
  URL for the Celery broker (e.g., Redis).

- **CELERY_RESULT_BACKEND**  
  Backend used by Celery to store task results.

## Environment Preferences

- **ENV_PROVIDER**  
  The environment provider(more coming soon). Options: `docker`.

- **DOCKER_RUNTIME**  
  Docker runtime to use. Example: `default`, you can use `sysbox-runc` if using Sysbox.

- **ENV_IMAGE**  
  Docker image to use for the environment. Example: `codercom/code-server:latest`, if using Sysbox use the docker image `ghcr.io/exoonix/enviro_dockerenv:latest`.

- **ENV_LIMITS**  
  Maximum number of environments a user can create. Set to `0` to disable the limit.

## Reverse Proxy Config

- **HOSTNAME**
  Set to the hostname of your application, for example: `onixtech.org`, only needed if using subdomain routing in `ROUTING_TYPE`.

- **ROUTING_TYPE**
  Set routing type. if using subpath, environments will be placed on /environments/<id>. If using subdomain, environments will be placed on env-<id>.example.com. You need `HOSTNAME` to be set. Options: `subpath`, `subdomain`.

- **SUBDOMAIN_PORTFORWARDING**
  This sets the envs to be forwarded to subdomains instead of subpaths. This is good for many frameworks like NextJS who want to be run on the base path. Values: `true`, `false`.

## Email Settings

- **EMAIL_BACKEND** Django email backend.  
    - For console: `django.core.mail.backends.console.EmailBackend`  
    - For SMTP: `django.core.mail.backends.smtp.EmailBackend`

- **EMAIL_HOST**  
  SMTP server host.

- **EMAIL_PORT**  
  SMTP server port. Default for SSL is `465`.

- **EMAIL_USE_TLS**  
  Set to `True` to use TLS.

- **EMAIL_USE_SSL**  
  Set to `True` to use SSL.

- **EMAIL_HOST_USER**  
  SMTP username.

- **EMAIL_HOST_PASSWORD**  
  SMTP password.

---

## Exporting Environment Templates to a Fixture

To export fixtures run:
```sh
./cli poetry run python manage.py dumpdata env_manager.EnvironmentTemplate --indent 2 > apps/env_manager/fixtures/templates.json
```

**Note:**  
After editing your `.env` file, restart your application for changes to take effect.

# Example Subpath Production Env

```
DJANGO_SECRET_KEY=your_secret_key
DEFAULT_SCHEME=https
DEBUG=False
DJANGO_LOGLEVEL=info
DJANGO_ALLOWED_HOSTS=localhost,django-docker,onixtech.org
DJANGO_CSRF_TRUSTED_ORIGINS=https://onixtech.org
DATABASE_ENGINE=postgresql_psycopg2
DATABASE_NAME=dockerdjango
DATABASE_USERNAME=dbuser
DATABASE_PASSWORD=dbpassword
DATABASE_HOST=db
DATABASE_PORT=5432

# Celery
CELERY_BROKER_URL=redis://redis:6379/0
CELERY_RESULT_BACKEND=redis://redis:6379/0

# Env preferences
ENV_PROVIDER = docker
DOCKER_RUNTIME = default
ENV_IMAGE = ghcr.io/exoonix/enviro_dockerenv:latest

ENV_LIMITS = 6


# Reverse proxy routing
HOSTNAME = onixtech.org # Only used if ROUTING_TYPE is subdomain
ROUTING_TYPE = subpath

SUBDOMAIN_PORTFORWARDING = false


# Email
EMAIL_BACKEND = django.core.mail.backends.console.EmailBackend # For smtp use django.core.mail.backends.smtp.EmailBackend


EMAIL_HOST = ''
EMAIL_PORT = 465
EMAIL_USE_TLS = False
EMAIL_USE_SSL = True

EMAIL_HOST_USER = ''
EMAIL_HOST_PASSWORD = ''
```

# Example Subdomain Production Env

```
DJANGO_SECRET_KEY=your_secret_key
DEFAULT_SCHEME=https
DEBUG=False
DJANGO_LOGLEVEL=info
DJANGO_ALLOWED_HOSTS=localhost,django-docker,onixtech.org,.onixtech.org
DJANGO_CSRF_TRUSTED_ORIGINS=https://onixtech.org
DATABASE_ENGINE=postgresql_psycopg2
DATABASE_NAME=dockerdjango
DATABASE_USERNAME=dbuser
DATABASE_PASSWORD=dbpassword
DATABASE_HOST=db
DATABASE_PORT=5432

# Celery
CELERY_BROKER_URL=redis://redis:6379/0
CELERY_RESULT_BACKEND=redis://redis:6379/0

# Env preferences
ENV_PROVIDER = docker
DOCKER_RUNTIME = default
ENV_IMAGE = ghcr.io/exoonix/enviro_dockerenv:latest

ENV_LIMITS = 6


# Reverse proxy routing
HOSTNAME = onixtech.org # Only used if ROUTING_TYPE is subdomain
ROUTING_TYPE = subdomain

SUBDOMAIN_PORTFORWARDING = true


# Email
EMAIL_BACKEND = django.core.mail.backends.console.EmailBackend # For smtp use django.core.mail.backends.smtp.EmailBackend


EMAIL_HOST = ''
EMAIL_PORT = 465
EMAIL_USE_TLS = False
EMAIL_USE_SSL = True

EMAIL_HOST_USER = ''
EMAIL_HOST_PASSWORD = ''
```

core/settings.py
"""
Django settings for core project.

Generated by 'django-admin startproject' using Django 5.2.1.

For more information on this file, see
https://docs.djangoproject.com/en/5.2/topics/settings/

For the full list of settings and their values, see
https://docs.djangoproject.com/en/5.2/ref/settings/
"""

from pathlib import Path
import os

# Build paths inside the project like this: BASE_DIR / 'subdir'.
BASE_DIR = Path(__file__).resolve().parent.parent


# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/5.2/howto/deployment/checklist/

# SECURITY WARNING: keep the secret key used in production secret!
SECRET_KEY = os.environ.get("DJANGO_SECRET_KEY")

# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = os.environ.get("DEBUG", "False").lower() == "true"

ALLOWED_HOSTS = os.environ.get("DJANGO_ALLOWED_HOSTS","127.0.0.1").split(",")
CSRF_TRUSTED_ORIGINS = os.environ.get("DJANGO_CSRF_TRUSTED_ORIGINS", "http://localhost:8081").split(",")

# ENVS
ENV_PROVIDER = os.environ.get("ENV_PROVIDER", "docker")
DOCKER_RUNTIME = os.environ.get("DOCKER_RUNTIME", "default")

ENV_IMAGE = os.environ.get("ENV_IMAGE", "codercom/code-server:latest")

ENV_LIMITS = int(os.environ.get("ENV_LIMITS", "0"))

DEFAULT_SCHEME = os.environ.get("DEFAULT_SCHEME", "http")

# Reverse proxy routing
HOSTNAME = os.environ.get("HOSTNAME", "onixtech.org")
ROUTING_TYPE = os.environ.get("ROUTING_TYPE", "subpath")

SUBDOMAIN_PORTFORWARDING = os.environ.get("SUBDOMAIN_PORTFORWARDING", "False").lower() == "true"

if ROUTING_TYPE == "subdomain" or SUBDOMAIN_PORTFORWARDING:
    SESSION_COOKIE_DOMAIN = f".{HOSTNAME}"
    CSRF_COOKIE_DOMAIN = f".{HOSTNAME}"

# Auth
AUTH_USER_MODEL = "users.CustomUser"
ACCOUNT_USER_MODEL_USERNAME_FIELD = None
ACCOUNT_SIGNUP_FIELDS = ['email*', 'password1*', 'password2*']
ACCOUNT_LOGIN_METHODS = {'email'}

LOGIN_REDIRECT_URL = '/dashboard/'


# Email verification
ACCOUNT_EMAIL_VERIFICATION = "mandatory"
ACCOUNT_EMAIL_VERIFICATION_BY_CODE_ENABLED = True

# Cache
CACHES = {
    "default": {
        "BACKEND": "django_redis.cache.RedisCache",
        "LOCATION": os.environ.get("CELERY_BROKER_URL", "redis://127.0.0.1:6379/1"),
        "OPTIONS": {
            "CLIENT_CLASS": "django_redis.client.DefaultClient",
        }
    }
}

AUTHENTICATION_BACKENDS = [
    # Needed to login by username in Django admin, regardless of `allauth`
    'django.contrib.auth.backends.ModelBackend',

    # `allauth` specific authentication methods, such as login by email
    'allauth.account.auth_backends.AuthenticationBackend',
]

# Application definition

INSTALLED_APPS = [
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    'django.contrib.humanize',
    # Celery
    'django_celery_beat',
    
    # Auth
    'allauth',
    'allauth.account',
    'allauth.usersessions',
    
    'allauth.socialaccount',
    
    'allauth.socialaccount.providers.github',
    # Htmx
    "django_htmx",

    'apps.users',
    # Apps
    'apps.home',
    "apps.env_manager",
    "apps.dashboard",
    "apps.blog",
    
    # Wagtail and debug
    'wagtail.contrib.forms',
    'wagtail.contrib.redirects',
    'wagtail.embeds',
    'wagtail.sites',
    'wagtail.users',
    'wagtail.snippets',
    'wagtail.documents',
    'wagtail.images',
    'wagtail.search',
    'wagtail.admin',
    'wagtail',

    'modelcluster',
    'taggit',
    "debug_toolbar",
]

MIDDLEWARE = [
    'django.middleware.security.SecurityMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
    
    # Whitenoise
    "django.middleware.security.SecurityMiddleware",
    "whitenoise.middleware.WhiteNoiseMiddleware",
    
    # Auth
    "allauth.account.middleware.AccountMiddleware",
    
    "django_htmx.middleware.HtmxMiddleware",
    "debug_toolbar.middleware.DebugToolbarMiddleware",
    'wagtail.contrib.redirects.middleware.RedirectMiddleware',
]

# Loading middleware
if ROUTING_TYPE == "subdomain" or SUBDOMAIN_PORTFORWARDING:
    MIDDLEWARE.append("apps.env_manager.middleware.host_check.HostnameMiddleware")

ROOT_URLCONF = 'core.urls'

TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [BASE_DIR / 'templates'],
        'APP_DIRS': True,
        'OPTIONS': {
            'context_processors': [
                'django.template.context_processors.request',
                'django.contrib.auth.context_processors.auth',
                'django.contrib.messages.context_processors.messages',
                
                # Allauth
                'django.template.context_processors.request',
            ],
        },
    },
]

WSGI_APPLICATION = 'core.wsgi.application'

STORAGES = {
    "default": {
        "BACKEND": "django.core.files.storage.FileSystemStorage",
    },
    "staticfiles": {
        "BACKEND": "whitenoise.storage.CompressedManifestStaticFilesStorage",
    },
}

# Database
# https://docs.djangoproject.com/en/5.2/ref/settings/#databases

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.{}'.format(
            os.getenv('DATABASE_ENGINE', 'sqlite3')
        ),
        'NAME': os.getenv('DATABASE_NAME', 'polls'),
        'USER': os.getenv('DATABASE_USERNAME', 'myprojectuser'),
        'PASSWORD': os.getenv('DATABASE_PASSWORD', 'password'),
        'HOST': os.getenv('DATABASE_HOST', '127.0.0.1'),
        'PORT': os.getenv('DATABASE_PORT', 5432),
    }
}


# Password validation
# https://docs.djangoproject.com/en/5.2/ref/settings/#auth-password-validators

AUTH_PASSWORD_VALIDATORS = [
    {
        'NAME': 'django.contrib.auth.password_validation.UserAttributeSimilarityValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.MinimumLengthValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.CommonPasswordValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.NumericPasswordValidator',
    },
]


# Internationalization
# https://docs.djangoproject.com/en/5.2/topics/i18n/

LANGUAGE_CODE = 'en-us'

TIME_ZONE = 'UTC'

USE_I18N = True

USE_TZ = True


# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/5.2/howto/static-files/
STATIC_ROOT = BASE_DIR / "staticfiles"
STATIC_URL = 'static/'
STATICFILES_DIRS = [
    BASE_DIR / "static",
]

MEDIA_ROOT = os.path.join(BASE_DIR, 'media')
MEDIA_URL = '/media/'

DATA_UPLOAD_MAX_NUMBER_FIELDS = 10_000

WAGTAIL_SITE_NAME = 'My Example Site'
WAGTAILADMIN_BASE_URL = 'http://example.com'
WAGTAILDOCS_EXTENSIONS = ['csv', 'docx', 'key', 'odt', 'pdf', 'pptx', 'rtf', 'txt', 'xlsx', 'zip']
# Default primary key field type
# https://docs.djangoproject.com/en/5.2/ref/settings/#default-auto-field

DEFAULT_AUTO_FIELD = 'django.db.models.BigAutoField'

# Email settings
EMAIL_BACKEND = os.getenv('EMAIL_BACKEND', 'django.core.mail.backends.smtp.EmailBackend')


EMAIL_HOST = os.getenv('EMAIL_HOST')
EMAIL_PORT = os.getenv('EMAIL_PORT')
EMAIL_USE_TLS = os.getenv('EMAIL_USE_TLS', 'False').lower() == 'true'
EMAIL_USE_SSL = os.getenv('EMAIL_USE_SSL', 'True').lower() == 'true'

EMAIL_HOST_USER = os.getenv('EMAIL_HOST_USER')
EMAIL_HOST_PASSWORD = os.getenv('EMAIL_HOST_PASSWORD')

DEFAULT_FROM_EMAIL = EMAIL_HOST_USER

INTERNAL_IPS = [
    "172.18.0.3",
    "127.0.0.1"
]

MESSAGE_STORAGE = 'django.contrib.messages.storage.session.SessionStorage'

# Celery

CELERY_BROKER_URL = os.environ.get("CELERY_BROKER_URL", "redis://127.0.0.1:6379/0")
CELERY_RESULT_BACKEND = os.environ.get("CELERY_RESULT_BACKEND", "redis://127.0.0.1:6379/0")
